Le drapeau de spam potentiel est disponible partout

AmauryCarrade opened the pull request:
Cette PR déplace le drapeau de spam potentiel (`is-potential-spam`) introduit par #5918 vers le modèle parent `Comment`, ainsi que la vérification des modifications de messages et le lever d'alerte. L'alerte est levée en respectant le bon `scope` en fonction du type de commentaire.

Une migration est ajoutée pour copier les données depuis le modèle `forum.Post` vers `utils.Comment`.

Une permission est également ajoutée pour donner le droit de voir et modifier le drapeau de spam potentiel.

Il reste des tests à ajouter, et peut-être à ajuster (on va voir ça très vite au statut de cette PR).

Closes #5962.

:warning: **La QA de cette PR nécessite de faire des actions _avant d'intégrer ces changements à votre instance locale_.**

### Fonctionnement

Le drapeau de spam potentiel a été déplacé vers `Comment`, et une API classique (sous `/api`) est désormais utilisée pour mettre ce drapeau à jour depuis l'interface en JS (comme le karma). Une route unique globale a été ajoutée, vu que ça ne change rien quelque soit le type de contenu.

La migration est un peu tricky : comme le champ de modèle d'origine n'existe plus en Python, du SQL pur a été utilisé pour lire les anciennes données, en renommant avant le champ, sinon un conflit se produit lors de la migration, Django interdisant d'avoir deux champs identiques dans un modèle et son modèle enfant — même lors de la migration. Les dépendances de ces migrations sont donc dans un ordre précis.

### Mise en production

Cette PR ajoute une permission pour voir et gérer les drapeaux de spam potentiel : `utils.change_comment_potential_spam`. Il faudra qu'un super-admin donne cette permission au groupe `staff`, et tout éventuel autre groupe approprié.

### Contrôle qualité

1. **Avant de passer sur la branche de cette PR dans votre instance locale**, marquez quelques messages du forum comme des spams potentiels.
2. Passez sur la branche de cette PR.
3. Exécutez les migrations (`make migrate-db`) et assurez-vous qu'aucune erreur ne remonte. Lancez le site avec `make run`.
4. Via l'administration de Django (en vous connectant avec `admin`/`admin`), [allez dans les paramètres du groupe `staff`](http://127.0.0.1:8000/admin/auth/group/1/change/) et ajoutez la permission `utils.change_comment_potential_spam`, autrement appelée _Can change the potential spam status of a comment_ (vous pouvez utiliser la recherche).
5. Retournez sur les messages du forum que vous avez marqué comme potentiel spam, et assurez-vous qu'ils le sont toujours, et que les autres ne le sont pas devenus.
6. En vous connectant dans un autre onglet avec le compte de l'auteur d'un message marqué comme potentiel spam, modifiez un de ses message du forum marqué comme tel, et assurez-vous qu'aucune erreur n'est remontée, que si l'auteur re-modifie à nouveau le message, l'alerte n'est pas dupliquée, et que l'auteur du message ne peut pas voir que le message est marqué comme spam potentiel (sauf si l'auteur est staff).
7. Faites de même avec un commentaire d'un tutoriel.
8. Faites de même avec un commentaire d'un article.
9. Faites de même avec un commentaire d'un billet.
10. Dans chaque cas, vérifiez que l'alerte dans le menu en haut à droite mène bien vers le commentaire.

AmauryCarrade committed:
Le drapeau de spam potentiel est disponible partout

Ce commit déplace le drapeau de spam potentiel (`is-potential-spam`)
vers le modèle parent Comment, ainsi que la vérification des modifica-
tions de messages et le lever d'alerte. L'alerte est levée en respec-
tant le bon `scope` en fonction du type de commentaire.

Une migration est ajoutée pour copier les données depuis le modèle
`forum.Post` vers `utils.Comment`.

Situphen commented on zds/utils/models.py:
```diff
@@ -442,6 +473,38 @@ def filter_usernames(original_list):
+                alert.save()
+            except Alert.DoesNotExist:
+                # We first have to compute the correct scope
+                sub_class = Comment.objects.get_subclass(id=self.id)
```
```suggestion
                sub_class = self.get_comment_subclass()
```

Situphen commented on zds/utils/models.py:
```diff
@@ -442,6 +473,38 @@ def filter_usernames(original_list):
+            except Alert.DoesNotExist:
+                # We first have to compute the correct scope
+                sub_class = Comment.objects.get_subclass(id=self.id)
+                if isinstance(sub_class, ContentReaction):
```
```suggestion
                if type(sub_class).__name__ == 'ContentReaction':
```

artragis replied:
pourquoi pas `isinstance`?

Situphen replied:
C'est simplement pour éviter l'importation locale de `ContentReaction` (voir le commentaire suivant), je trouve ça plus propre (mais c'est subjectif)

Situphen commented on zds/utils/models.py:
```diff
@@ -442,6 +473,38 @@ def filter_usernames(original_list):
+        :param old_text: old version of this message.
+        :param new_text: new version of this message.
+        """
+        # Local import to avoid circular import as ContentReaction inherits from Comment.
+        from zds.tutorialv2.models.database import ContentReaction
```
```suggestion
```

AmauryCarrade committed:
Évite un import (suggestion de Situphen)

Co-authored-by: Situphen <Situphen@users.noreply.github.com>

AmauryCarrade committed:
Utilise une méthode dédiée pour récupérer la sous-classe de Comment

Co-authored-by: Situphen <Situphen@users.noreply.github.com>

AmauryCarrade committed:
Lint

AmauryCarrade committed:
Utilisation de signaux pour envoyer les alertes.

Auparavant, les alertes étaient envoyées depuis la méthode
`update_content` du modèle `utils.Comment`. Le problème, c'est que lors
de l'appel de cette méthode, l'éditeur n'est généralement pas modifié,
ce qui rend son comportement incohérent car l'alerte n'est envoyée que
si l'auteur lui-même modifie le message. Un signal `post_save` est donc
utilisé pour envoyer l'alerte.

AmauryCarrade committed:
Mise à jour des tests pour prendre tous les cas en compte.

Des tests ont été ajoutés pour les contenus, et les tests du forum ont
été ajustés. De plus, les parties communes des tests forum et contenus
ont été factorisées.

AmauryCarrade commented:
Comme mentionné dans ba60acb086657e07c891c192a3a7d2c48b2663aa, j'ai dû restructurer la façon dont ceci est implémenté.

Auparavant, les alertes étaient envoyées depuis la méthode `update_content` du modèle `utils.Comment`. Le problème, c'est que lors de l'appel de cette méthode, l'éditeur n'est généralement pas modifié, ce qui rend son comportement incohérent car l'alerte n'est envoyée que si l'auteur lui-même modifie le message. Un signal `post_save` est donc utilisé pour envoyer l'alerte. Le code change un peu pour s'adapter à ce nouveau contexte.

Ce signal ne spécifie pas d'expéditeur (`@receiver(post_save, sender=…)`), car il faut attraper les `Comment` et ses classes filles ; or, le système de Django vérifie que l'expéditeur est exactement le même que le filtre donné, sans vérifier l'héritage si c'est une classe. À la place, un test est effectué dans la fonction du signal. Je ne pense pas que ça ait d'implication, mais si quelqu'un connaissant mieux Django avancé que moi pense que ça peut poser des soucis (p.ex. de performances) et/ou a une méthode alternative à proposer (qui ne consiste pas en mettant trois `@receiver` pour chaque classe, car trop gros risque d'oubli en cas d'ajout d'une sous-classe de `Comment`), je suis preneur.

Petite optimisation apportée par cette PR au passage : `Comment.update_content()` **ne sauvegarde plus l'instance**, car c'était de toute façon systématiquement fait après lors de l'usage de cette méthode. Ça évite des requêtes inutiles. Ce point a été ajouté à sa documentation.

Sinon, cette PR est prête à être testée.

Situphen commented:
> Auparavant, les alertes étaient envoyées depuis la méthode `update_content` du modèle `utils.Comment`. Le problème, c'est que lors de l'appel de cette méthode, l'éditeur n'est généralement pas modifié, ce qui rend son comportement incohérent car l'alerte n'est envoyée que si l'auteur lui-même modifie le message. Un signal `post_save` est donc utilisé pour envoyer l'alerte. Le code change un peu pour s'adapter à ce nouveau contexte.

- Bob écrit un message sur un sujet
- Alice, modératrice, modifie ce message
- Alice, modératrice, marque ce message comme spam potentiel
- Bob, plusieurs mois après, modifie son message
- Actuellement, aucune alerte n'est levée car lors de l'appel à `Comment.update_content()` `Comment.editor` n'a pas encore été modifié et pointe toujours vers Alice. En utilisant le signal `post_save`, l'alerte est levée car `Comment.editor` a été modifié et pointe vers Bob.

J'ai bien compris ?

Je ne vois pas vraiment de situation réaliste où un Staff irait modifier le contenu d'un message _et_ le marquer comme spam potentiel, mais il est vrai que c'est un comportement incohérent (un *edge case* très fourbe) !

AmauryCarrade commented:
Oui, mais c'est aussi vrai (et beaucoup moins rare !) à la première modification d'un message, car au moment de l'appel à `update_content`, `self.editor` vaut généralement toujours `None`. Ces deux cas étaient de toute façon testés, et faisaient échouer les tests.

Plus généralement, `editor` avait pour valeur l'éditeur précédent, et non courant, au moment de l'appel de la méthode. Le troisième point de ta liste est d'ailleurs optionnel (sinon qu'il faut certes que le message soit marqué). Et le second aussi, en regard du paragraphe précédent.

Laisser cette méthode s'occuper de ça faisait un contrat implicite — `editor` devrait être mis à jour avant d'appeler `update_content` — ce qui n'est selon moi pas acceptable (même documenté) tant c'est un nid à bugs futurs. D'où la solution du signal. Une réécriture de `save` pourrait éventuellement être envisageable, aussi.

----

Un test sur les pings échoue tout d'un coup, faudra que je regarde pourquoi… Là comme ça je ne vois pas mais j'imagine qu'il doit y avoir un truc fourbe dans un coin.

AmauryCarrade committed:
Removed backup file

AmauryCarrade committed:
Restaure les pings comme avant

AmauryCarrade committed:
Vraiment comme avant

AmauryCarrade committed:
Oh en fait la sauvegarde dans `update_content` était nécessaire

AmauryCarrade commented:
Cette PR est prête.

Situphen committed:
Merge branch 'dev' into un-signalement-pour-les-surveiller-tous-et-dans-les-tenebres-les-detecter

Situphen commented on zds/utils/models.py:
```diff
@@ -413,13 +416,36 @@ class Meta:
+        This method will render the new comment to HTML, store the rendered
+        version, and analyze pings.
+
+        This method updates fields and pings, but won't save the instance.
```
```suggestion
        This method updates fields and pings, and save the instance as well.
```

Situphen commented on zds/utils/tests/tests_comments.py:
```diff
@@ -0,0 +1,197 @@
+        self.assertEqual(response.status_code, status.HTTP_302_FOUND)
+        self.assertEqual(len(Alert.objects.filter(author=bot, comment=comment, text=alert_text, solved=False)), 0)
+
+        # authenticated as non-staff, if we edit the message, there is no alert
```
```suggestion
        # authenticated as staff, if we edit the message, there is no alert
```

AmauryCarrade committed:
Forgot to update comment after method behaviour change

Co-authored-by: Situphen <Situphen@users.noreply.github.com>

AmauryCarrade committed:
Misleading test comment

This survived the previous code review, oups

Co-authored-by: Situphen <Situphen@users.noreply.github.com>

Situphen reviewed (changes requested):
**QA NOK** Cas à reproduire qui pose soucis

- `user` crée un message
- `user` édite son message
- `admin` marque ce message comme spam potentiel
- `admin` édite le message en question

Ce qui se produit : une alerte est levée

Ce qui devrait se produire : aucune alerte n'est levée

-----

Autre cas pour aider à dépister le bogue

- `user` crée un message
- `admin` marque ce message comme spam potentiel
- `admin` édite le message en question
- Pas d'alerte
- `admin` édite le message en question
- Pas d'alerte
- `admin` enlève le marquage comme spam potentiel
- `user` édite le message
- `admin` marque ce message comme spam potentiel
- `admin` édite le message en question
- Une alerte est levée

Le soucis vient donc quand `admin` édite le message juste après une édition de `user`

AmauryCarrade commented:
Je recopie ce que je disais en privé

En gros là le cycle d'enregistrement c'est généralement :

1. appel à `update_content` (+ `save()` dans un état inconsistant : le contenu a changé mais pas le reste des métadonnées) ;
2. modification des autres métadonnées (dont le modificateur du message) ;
3. second appel à `save()` depuis le code modifiant le message.

Le signal est appelé à chaque `save`, donc au premier, l'éditeur est encore l'ancien éditeur. Le souci c'est que les pings ont besoin de la sauvegarde, car sinon potentiellement le message n'est pas enregistré (à la création) et c'est nécessaire pour créer les notifications.

Ma solution, c'est de déplacer l'analyse des pings ou dans un signal, ou dans la méthode `save()` surchargée, comme ça plus de `save()` dans `update_content`. Et un seul `save()` par enregistrement, évitant les enregistrements dans un état inconsistants.

Et d'ajouter un test pour surveiller ce cas spécifique.

AmauryCarrade committed:
On n'enregistre plus Comment dans un état incohérent

Ce commit évite d'enregistrer des modèles Comment dans un état
inconsistant dans la base de données. Auparavant, les messages étaient
généralement enregistrés deux fois : une fois avec juste le contenu
mis à jour (mais pas les métadonnées : à ce stade, l'enregistrement est
donc incohérent avec les données réelles), puis une autre avec cette
fois les métadonnées (à ce stade, c'est donc bon).

Le souci, c'est que l'analyse de l'antispam, effectuée automatiquement
à l'enregistrement, a besoin de données cohérentes (à savoir le nouveau
contenu et l'éditeur du message). Cet état de fait faisait donc des
alertes parasites ou manquantes dans certains cas.

Ce commit corrige ce comportement en s'assurant qu'un Comment n'est
jamais enregistré en base de données dans un état incohérent, et en
déplaçant l'analyse des pings et celle de l'anti-spam dans la méthode
`save`.

Au passage, les performances sont très légèrement améliorées, étant donné
qu'un certain nombre de requêtes inutiles sont évitées.

AmauryCarrade commented:
Le souci remonté par la QA a été corrigé, et un test a été ajouté afin qu'il ne se reproduise plus.

Situphen commented on zds/utils/tests/tests_comments.py:
```diff
@@ -0,0 +1,213 @@
+        self.assertEqual(response.status_code, status.HTTP_302_FOUND)
+        self.assertEqual(len(Alert.objects.filter(author=bot, comment=comment, text=alert_text, solved=False)), 0)
+
+        # Now we test a tricky scenarii: an user create a message, then edit te message; then the message is marked as
+        # span and _a staff_ edit the message. No alert should be sent.
+        # The message was already modified by its author just before, so we'll start at the second step.
```
```suggestion
        # Now we test a tricky scenario:
        # 1. a user creates a message;
        # 2. the same user edits the message;
        # 3. a staff member marks the message as potential spam.
        # No alert should be sent.
        # The message was already modified by its author just before, so we'll start at the third step.
```

AmauryCarrade replied:
Avec ta suggestion ce serait `third step` du coup (dernière ligne)

Si tu corriges la suggestion je l'applique !

Situphen replied:
Done.

Situphen commented on zds/utils/tests/tests_comments.py:
```diff
@@ -0,0 +1,213 @@
+        self.assertEqual(response.status_code, status.HTTP_302_FOUND)
+        self.assertEqual(len(Alert.objects.filter(author=bot, comment=comment, text=alert_text, solved=False)), 0)
+
+        # if the author edit the message, there is no alert either
```
```suggestion
        # if the author edits the message, there is no alert either
```

Situphen commented on zds/utils/tests/tests_comments.py:
```diff
@@ -0,0 +1,213 @@
+        for alert in Alert.objects.filter(author=bot, comment=comment, text=alert_text, solved=False):
+            alert.solve(staff.user)
+
+        # if the staff edit the message, there is no alert
```
```suggestion
        # if the staff edits the message, there is no alert
```

Situphen commented on zds/utils/tests/tests_comments.py:
```diff
@@ -0,0 +1,213 @@
+        self.assertEqual(response.status_code, status.HTTP_302_FOUND)
+        self.assertEqual(len(Alert.objects.filter(author=bot, comment=comment, text=alert_text, solved=False)), 1)
+
+        # authenticated as non-staff, if we edit again the message, there is still one alert (we dont stack them up)
```
```suggestion
        # authenticated as non-staff, if we edit again the message, there is still one alert (we don't stack them up)
```

Situphen commented on zds/utils/models.py:
```diff
@@ -413,12 +416,66 @@ class Meta:
+                         See render_markdown.
+        """
+        # This attribute will be used by `_save_check_spam`, called after save (but not saved into the database).
+        # We only update it if it does not already exist, so if this method is called multiple times, the oldest
```
```suggestion
        # We only update it if it does not already exists, so if this method is called multiple times, the oldest
```

Situphen commented on zds/utils/api/permissions.py:
```diff
@@ -0,0 +1,10 @@
+
+class UpdatePotentialSpamPermission(permissions.BasePermission):
+    """
+    Global permission check for blocked IPs.
```
Cette description ne me parait pas adéquate. Je ne vois pas pourquoi ça parle d'IPs bloquées.

AmauryCarrade replied:
J'avoue que je n'ai aucune idée de pourquoi cette description.
```suggestion
    Permission to use the API to mark a comment as (non)spam.
```

AmauryCarrade committed:
Apply suggestions from code review

Co-authored-by: Situphen <Situphen@users.noreply.github.com>

AmauryCarrade committed:
Update misleading documentation

Situphen commented:
**QA OK** Tout est bon pour moi !
